### ISSUE TITLE
ENH: Transforms and Normalizations

### ISSUE THREAD

**Issue by jbrockmendel** [statsmodels/statsmodels#3627]
Here's an item for the wish-list.  Consider for example a VAR with unemployment as one of the variables.

```
>>> data = sm.datasets.macrodata.load_pandas()
>>> df = data.data
>>> model = sm.tsa.VAR(df[['realgdp', 'unemp', 'realint']])
>>> res = model.fit(4)
>>> fc = res.forecast_interval(res.endog, 20, alpha=.00001)
# Note: even if we can't agree on how to attach dates, we should at least attach variable names here
>>> fc[1][-1][1] # lower-bound of confidence interval 20 periods out
-0.6053041064870035
```
Note that using newer data doesn't require such an extreme `alpha` to make this happen.  Negative unemployment isn't very helpful, so we try again.  One approach is to transform the unemployment column to have a domain along the entire real line:

```
>>> unemp01 =  df['unemp']/100.
>>> df['unemp2'] = -np.log(1/unemp01 - 1)
>>> model = sm.tsa.VAR(df[['realgdp', 'unemp2', 'realint']])
>>> res = model.fit(4)
>>> fc = res.forecast_interval(res.endog, 20, alpha=.00001)
>>> for n in [0, 1, 2]:
    # Apply the inverse transform
    fc[n][:, 1] = 100/(1+np.exp(-fc[n][:, 1]))

>>> fc[1][-1][1]
1.8821543543947881
```

So the more general idea I'm suggesting is handling situations where the ultimate outputs should be in the same units as the user inputs, but where some intermediate handling is necessary.  The two big categories of cases that come to mind are 1) non-negative variables and 2) integrated variables.

**Comment by josef-pkt** [comment]
This was the idea behind adding boxcox #3477, also used for forecasting in tbats #2892 
issue #2791 is a more general case

forecast interval could be based on end-point transformation like you have it, forecasting needs an option similar to ARIMA for `typ='level'` when integrated.

For VAR it's less clear to me because the impulse response function and other AR-process features rely on the linear process, maybe inverse transforming whatever the outcome is would work. However, we  have expectation in terms of the transformed and not the original endog.

API question: It wasn't clear to me how we get this information from the user (or from the formula). Most likely we need a keyword for the model where the user specifies the transformation.

**Comment by jbrockmendel** [comment]
I'll take a look at those, thanks.

> For VAR it's less clear to me because the impulse response function and other AR-process features rely on the linear process, maybe inverse transforming whatever the outcome is would work. 

It might make sense to treat this like the various `data.attach_foo` methods to only un-transform at the end-output stage, if at all.

> API question: It wasn't clear to me how we get this information from the user (or from the formula).  Most likely we need a keyword for the model where the user specifies the transformation.

If a user is willing to trust sensible defaults, then in many cases specifying bounds is enough. `bounds={'unemp': (0, 1), 'age': (0, None)}` gets us most of the way there.  (Though we'd still need to be careful about open/closed endpoints).

**Comment by josef-pkt** [comment]
> If a user is willing to trust sensible defaults, then in many cases specifying bounds is enough. bounds={'unemp': (0, 1), 'age': (0, None)} gets us most of the way there.

No, I don't think we would want automatic choices except in convenience models. The transformation changes the endog and the implied distribution, e.g. log(endog) imposes log-normal distribution with the usual normal distribution interpretation of linear models. boxcox would be a more general, parameterized alternative. For endog in a fixed interval like (0, 1) we have many choices among the GLM link functions (by analogy).
Users would have to decide which transformation they find acceptable.

As partial support to get started, I have a PR, IIRC, to add nonlinear prediction/test of restrictions including confidence interval, which would be a new method to implement the main consequences transformation.

(Most transformation for bounds when applied to endog assume that the interval is open, i.e. boundary values are excluded, like log or logit. This is less restrictive in GLM because there we apply the inverse link transformation on the linear predictor and we don't have restrictions on endog to exclude boundaries. e.g. instead of log(endog) we use exp(lin_pred) as in proisson.)

**Comment by josef-pkt** [comment]
(As connection to boxcox: The transformation affects the distribution and variance function. This is intentionally used for the choice of the boxcox parameter when we want to stabilize the variance. Just for imposing bounds the change in the variance function might not be desirable or appropriate.)

**Comment by jbrockmendel** [comment]
> No, I don't think we would want automatic choices except in convenience models. 

Fair enough.  Out-of-scope is a pretty reasonable default.

**Comment by josef-pkt** [comment]
I didn't say out-of-scope.

It might be out-of-scope for the core parts of the models (to avoid bloating the core computation), but
I think the use case is relevant and, for example, prediction and similar should eventually support it. The question is how to support it without making the models too big, and how users should specify what they want.
(It's the joy of having to come up with a good or reasonable design. :)
